[ZIP 216] Require Canonical Jubjub Point Encodings (#442)
zcash/zips: zip-0216.rst

  anywhere that a prime-order subgroup point is specified.
- `(0, 1)`, which is the identity, and thus is an element of the prime-order subgroup.

The non-canonical identity creates a consensus issue because unlike other non-canonical
points that are rejected, a non-canonical identity that is decoded and then encoded, does
not produce the original encoding. For example, if a non-canonical encoding appeared in a
transaction field, then node implementations that store points internally as abstract
curve points, and used those to derive transaction IDs, would derive different IDs than
nodes which store transactions as bytes (such as `zcashd`).

Specification
=============

TODO: Define a non-canonical Jubjub point encoding.

When this ZIP activates, all places within the Sapling consensus protocol where Jubjub
points occur MUST reject non-canonical Jubjub point encodings.

- Sapling addresses:
  - `pk_d`
- Sapling Spend description:
  - `cv`
  - `rk`
  - The `R` component of the `spendAuthSig` RedDSA signature.
- Sapling Output description:
  - `cv`
  - `ephemeralKey`
- The `R` component of the `bindingSigSapling` RedDSA signature.

> 

Reviewer: ```suggestion
TODO: specify when addresses, full viewing keys, and note plaintexts are checked.

```

Rationale
=========

Zcash previously had a similar issue with non-canonical representations of points in
Ed25519 public keys and signatures. In that case, given the prevalence of Ed25519
signatures in the wider ecosystem, the decision was made in ZIP 215 [#zip-0215]_ (which
activated with the Canopy network upgrade [#zip-0251]_) to allow non-canonical
representations of points.

> Here, we are motivated instead to reject these non-canonical points:

Reviewer: ```suggestion
In Sapling, we are motivated instead to reject these non-canonical points:
```

> - The chance of the identity occuring anywhere within Sapling transactions from
>   implementations following the standard protocol is cryptographically negligible.

Reviewer: ```suggestion
- The chance of the identity occurring anywhere within the Sapling components of
  transactions from implementations following the standard protocol is cryptographically
  negligible.
```

- This re-enables the aforementioned simpler security analysis of the Sapling protocol.

> - The Jubjub curve has a vastly-smaller scope of usage than Curve25519.

Reviewer: ```suggestion
- The Jubjub curve has a vastly-smaller scope of usage in the general cryptographic
  ecosystem than Curve25519 and Ed25519.
```

Security and Privacy Considerations
===================================

This ZIP eliminates a potential source of consensus divergence between differing full node
implementations. At the time of writing (February 2021), no such divergence exists for any
production implementation of Zcash, but the alpha-stage `zebrad` node implementation would
be susceptible to this issue.

Deployment
==========

> TBD

Reviewer: ```suggestion
This ZIP is proposed to activate with Network Upgrade 5.
```

References
==========

.. [#RFC2119] `RFC 2119: Key words for use in RFCs to Indicate Requirement Levels <https://www.rfc-editor.org/rfc/rfc2119.html>`_
.. [#protocol-jubjub] `Zcash Protocol Specification, Version 2020.1.15. Section 5.4.8.3: Jubjub <protocol/protocol.pdf#jubjub>`_

> .. [#jubjub-crate] `jubjub <https://crates.io/crates/jubjub>`_

Reviewer: ```suggestion
.. [#jubjub-crate] `jubjub Rust crate <https://crates.io/crates/jubjub>`_
```
